tools/c7n_azure/c7n_azure/resources/network_security_group.py

```diff
@@ -0,0 +1,207 @@
+        return network_security_groups
+
+    def is_range_match(self, dest_port_range):
+        # destination port range is coming from Azure, existing rules, not policy input
+        if len(dest_port_range) > 2:
+            raise ValueError('Invalid range')
+        # Both FromPort and ToPort are specified, should fall within range
+        if self.from_port and self.to_port:
+            for port in dest_port_range:
+                if port > self.to_port or port < self.from_port:
+                    return False
+        # Just FromPort is specified, should be above FromPort
+        elif self.from_port:
+            for port in dest_port_range:
+                if port < self.from_port:
+                    return False
+        # Just ToPort is specified, should be below ToPort
+        elif self.to_port:
+            for port in dest_port_range:
+                if port > self.to:
+                    return False
+        # OnlyPorts is specified, anything NOT included in OnlyPorts should return True
+        elif self.only_ports:
+            for op in self.only_ports:
+                if len(dest_port_range) > 1:
+                    if dest_port_range[0] <= op >= dest_port_range[1]:
+                        return False
+                else:
+                    if dest_port_range[0] == op:
+                        return False
+        # Ports is specified, only those included in Ports should return true
+        elif self.ports:
+            if len(dest_port_range) > 1:
+                # self.ports needs to have ALL ports in range (inclusive) to match
+                for i in range(dest_port_range[0], dest_port_range[1] + 1):
+                    if i not in self.ports:
+                        return False
+            else:
+                if dest_port_range[0] not in self.ports:
+                    return False
+        return True
+
+    def is_ranges_match(self, security_rule):
+        if 'destinationPortRange' in security_rule['properties']:
+            dest_port_ranges = \
+                [self.get_port_range(security_rule['properties']['destinationPortRange'])]
+        else:
+            dest_port_ranges = \
+                [self.get_port_range(range_str) for range_str in security_rule['properties']['destinationPortRanges']]
+        for range in dest_port_ranges:
+            if not self.is_range_match(range):
+                return False
+        return True
+
+    def get_port_range(self, range_str):
+        return [int(item) for item in range_str.split('-')]
+
+    '''
+    Determine if SecurityRule matches criteria as entered in policy
+
+    Currently supporting filters:
+        Ports
+        OnlyPorts
+        FromPort
+        ToPort
+        IpProtocol
+    '''
+
+    def is_match(self, security_rule, match_op):
+        direction_match = self.direction_key == security_rule['properties']['direction']
```

Comment:
it would be better to return early here if the bool eval is False as the last return statement requires this to be True to prevent additional checks in is_ranges_match/is_range_match.

```python
if self.direction_key != security_rule['properties']['direction']:
  return False
```